### ISSUE TITLE
Raw values stringified

### ISSUE THREAD

**Issue by betocantu93** [hjdivad/ember-m3#807]
Hello, I'm working in a PoC using ember-m3 and DS.Models for a possible incremental migration, it's going great so far, so many unlocked possibilities! I'm using the current `master` branch for the recently added computeAttribute API, but I'm getting a weird error... sometimes arrays and/or objects raw values get stringified  and I'm unable to replicate it consistently yet, any suggestions? thanks in advance

![Screen Shot 2020-07-20 at 19 30 05](https://user-images.githubusercontent.com/9092644/88005072-10135e80-cace-11ea-9fb8-9f676398cc04.png)

Here is a reproduction

https://github.com/betocantu93/show-stringified-bug

If you open ember inspector/data tab select first ref:document, get the instance into $E and...

try: `$E.get('prop1.table)` or : `$E.get('prop1.chair)` where prop1 is a nested object

you will get the table array stringified or chair object stringified

I'm not sure why, but this happens with app reloads

**Comment by betocantu93** [comment]
It seems that its always stringified unless you reload the app with the ember inspector/data tab opened with the faulty model type selected, like if its lazy evaluated or something, gets stringified?

**Comment by betocantu93** [comment]
I think its a bug with the ember inspector, if you reload in the data tab with a **Megamorphic** type selected, it gets this weird bug.

**Comment by betocantu93** [comment]
I just found out that if you kept a megamorphic type selected on the ember inspector _data_ tab between app reloads other megamorphic types get their arrays and object raw values stringified, I opened an issue here: https://github.com/hjdivad/ember-m3/issues/807

It's kind of weird, the literal workaround is to just don't keep a megamorphic type selected between reloads...

**Comment by SYU15** [comment]
@betocantu93 thanks for flagging this. I can't repro on the m3 dummy app however, I am fairly confident the issue is probably related to [these lines of code](https://github.com/hjdivad/ember-m3/blob/master/addon/adapters/m3-debug-adapter.js#L64-L68) where we stringify the object value so that it doesn't render as `[object object]` in the inspector. We put this in place due to the constraints of Ember Inspector which does not traverse the data structures deep enough for performance reasons iirc to display in the inspector table.

So this is ultimately in place because the table view as it is now is not the ideal way of displaying this type of data and we need to put a workaround in place.

So I guess there are 2 options (both are possible):
1. Fix the issue by potentially re-parseing the data before it gets sent to the console (I admit this is not an ideal solution lmk if anyone can think of another). We could probably do something [here](https://github.com/emberjs/ember-inspector/blob/50db91b7bd26b12098cae774a307208fe0a47d75/ember_debug/data-debug.js#L185) maybe?
2. Spend some time designing a new pane in the inspector that better suits deeply nested data as I have discussed briefly with @igorT (the vision for this is fairly TBD, correct me if I am wrong).

Let me know what you think!

cc @chancancode

**Comment by SYU15** [comment]
Ok I thought of another solution @betocantu93  @igorT. We can revert the `stringify` change and instead have nested objects/arrays rendered as clickable `...` buttons that sends the nested object to the right-hand pane to be inspected similarly to what we did for arguments in the component tree: https://github.com/emberjs/ember-inspector/pull/1175

One tradeoff is that would mean if the user clicked on the `$E`, it would only send the nested object to the console, and not the top level record, which is correct/consistent behavior but may confuse end users who may not be expecting that/are familiar with the old flow.

Also this change would not only impact `m3` but any other data library that hooks into the inspector via a `data-adapter` so it should be treated with care.

**Comment by betocantu93** [comment]
@SYU15 thanks for the follow-up

> 1. Fix the issue by potentially re-parseing the data before it gets sent to the console (I admit this is not an ideal solution lmk if anyone can think of another). We could probably do something [here](https://github.com/emberjs/ember-inspector/blob/50db91b7bd26b12098cae774a307208fe0a47d75/ember_debug/data-debug.js#L185) maybe?

I think I would need more context to properly discard this approach but, it seems that its being mutated by reference (arrays and objects) even without interacting with the "send to console" feature, the bug exists. Maybe a stringified clone may "hack" this for tables? parsing when sending to the console would not be enough, for instance

```hbs
{{#each object.nested.nested as |nestedNestedItem|}}
```
won't work, bc object.nested.nested is an string at this point when the bug appears


> Ok I thought of another solution @betocantu93 @igorT. We can revert the `stringify` change and instead have nested objects/arrays rendered as clickable `...` buttons that sends the nested object to the right-hand pane to be inspected similarly to what we did for arguments in the component tree: [emberjs/ember-inspector#1175](https://github.com/emberjs/ember-inspector/pull/1175)

I think this approach sounds interesting, and makes a lot of sense to me...


> Also this change would not only impact `m3` but any other data library that hooks into the inspector via a `data-adapter` so it should be treated with care.

For m3 records, each nested model have a reference `this._parentModel` so maybe m3 users won't really have a hard to traverse up if needed? A big part of my m3 usage are nested models so maybe I'm biased because this actually would facilitate stuff for my scenario, like I don't know how non-nested models (complex POJOS) would work out... maybe wrapping the clicked object and displaying a navigate to parent button in the pane

```ts
objectClicked;
actualSentObject = {
  parent,
  objectClicked
};
```

I don't really have context in the integration between inspector and data-adapter, but maybe this could be something under a flag that m3 or other libs use? like show `OPEN_NESTED_OBJECTS_IN_PAGE`

**Comment by SYU15** [comment]
> I think I would need more context to properly discard this approach but, it seems that its being mutated by reference (arrays and objects) even without interacting with the "send to console" feature, the bug exists. Maybe a stringified clone may "hack" this for tables? parsing when sending to the console would not be enough, for instance

Yeah just to be clear this isn't a "true" fix; the data is already mutated even before it gets sent to Ember Inspector, and if we parse it before it gets sent to the console, we will probably reintroduce the bug this code was put in place to fix. I guess we could stringify it somewhere in the inspector code just for the table instead but...meh. It isn't a great solution to begin with.

In my mind, the cleanest option would be the `...` button, but perhaps we can discuss over discord or during the data meeting.

**Comment by betocantu93** [comment]
I see... I also think the `...` approach is probably the way to go, if I can help, please let me know, happy to jump in to any conversation or meeting.

**Comment by igorT** [comment]
Another lightweight way of fixing this would be for m3 to instead of passing a stringified object pass a a megamorphicModel, that has been wrapped behind a proxy which proxies the array properties to return arrays which have a more sensible `toString` method. This is a bit bonkers, but might be a quick hack to make this work before the longer term solution

**Comment by SYU15** [comment]
Hi @betocantu93 are you interested in taking the initiative and investigating either my or @igorT's solution? I am happy to discuss further or review PRs.

**Comment by betocantu93** [comment]
@SYU15 yes sure, I can help with this one, I will find some time this weekend, not sure which approach yet, I'll investigate 👍🏼 thanks

**Comment by SYU15** [comment]
Great! Feel free to reach out next week or so.

**Comment by SYU15** [comment]
Also, to clarify, it is possible to do both fixes (first the shorter term fix @igorT suggested and then mine) in a phased approach if we agree on what problem this fix is trying to address. But I will leave that up to you @betocantu93 to explore.

**Comment by betocantu93** [comment]
I agree with the phases, the scope of this fix should be to address the bug for now, not necessarily the ergonomics, we can start to plan for that too, happy to get involved

**Comment by betocantu93** [comment]
Closed by #817
